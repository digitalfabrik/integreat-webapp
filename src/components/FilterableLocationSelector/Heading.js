import React from 'react'

import LocationBig from './assets/LocationBig.svg'
import style from './Heading.css'
import { translate } from 'react-i18next'
import Caption from 'components/UIComponents/Caption'

class Heading extends React.Component {
  render () {
    const {t} = this.props
    return (
      <div>
        <img className={style.logo} src={LocationBig}/>
        <Caption className={style.caption} title={t('Location:where')}/>
      </div>
    )
  }
}

export default translate('Location')(Heading)
