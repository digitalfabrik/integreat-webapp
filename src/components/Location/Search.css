:root {
  --search-logo-width: 20px;
}

.search {
  width: 80%;
  margin: auto;
}

.search-icon {
  width: var(--search-logo-width);
  font-size: 1.2em;
  text-align: center;
}

.search-input {
  width: calc(100% - var(--search-logo-width) - 5px);
  height: 25px;
  margin-left: 5px;

  color: black;
  background: transparent;

  border-width: 0 0 1px;
  border-color: black;
  outline: none;
  border-radius: 0;

  &::placeholder {
    color: black;
  }
}
