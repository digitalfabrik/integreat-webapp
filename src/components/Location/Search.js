import React from 'react'
import PropTypes from 'prop-types'
import FontAwesome from 'react-fontawesome'
import { translate } from 'react-i18next'

import style from './Search.css'

class Search extends React.Component {
  static propTypes = {
    filterText: PropTypes.string,
    onFilterTextChange: PropTypes.any
  }

  render () {
    const {t} = this.props
    return (
      <div>
        <div className={style.search}>
          <FontAwesome className={style.searchIcon} name='search'/>
          <input type='text' placeholder={t('Location:search')} className={style.searchInput} defaultValue={this.props.filterText}
                 onChange={(event) => this.props.onFilterTextChange(event.target.value)}/>
        </div>
      </div>
    )
  }
}

export default translate('Location')(Search)
