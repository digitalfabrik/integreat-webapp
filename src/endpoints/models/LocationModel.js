import { isEmpty } from 'lodash/lang'
import { find } from 'lodash/collection'

const IGNORED_PREFIXES = ['stadt', 'kreis', 'landkreis']

export default class LocationModel {
  constructor (name, code, live) {
    this._name = name
    this._code = code
    this._live = live
    this.initializeSorting()
  }

  get live () {
    return this._live
  }

  get name () {
    return this._name
  }

  get code () {
    return this._code
  }

  get sortKey () {
    return this._sortKey
  }

  get sortCategory () {
    return this._sortCategory
  }

  initializeSorting () {
    let name = this._name.toLowerCase()
    const prefix = find(IGNORED_PREFIXES, (pre) => name.startsWith(pre + ' '))

    if (prefix) {
      name = name.substring(prefix.length + 1).trim()
    }

    this._sortKey = name
    this._sortCategory = isEmpty(name) ? '?' : name[0].toUpperCase()
  }
}
