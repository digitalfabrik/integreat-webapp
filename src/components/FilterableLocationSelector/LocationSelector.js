import React from 'react'
import PropTypes from 'prop-types'

import style from './style.css'
import { transform } from 'lodash/object'
import { Link } from 'redux-little-router'
import { groupBy, filter } from 'lodash/collection'
import LocationModel from '../../endpoints/models/LocationModel'

class LocationParentEntry extends React.Component {
  static propTypes = {
    name: PropTypes.string.isRequired
  }

  render () {
    return (
      <div className={style.languageListParent}>{this.props.name}</div>
    )
  }
}

class LocationEntry extends React.Component {
  static propTypes = {
    language: PropTypes.string,
    location: PropTypes.object.isRequired
  }

  render () {
    const location = this.props.location
    return (
      <Link href={`/${location.code}/${this.props.language}`} className={style.languageListItem}>{location.name}</Link>
    )
  }
}

class LocationSelector extends React.Component {
  static propTypes = {
    locations: PropTypes.arrayOf(PropTypes.instanceOf(LocationModel)),
    filterText: PropTypes.string.isRequired,
    language: PropTypes.string
  }

  filter () {
    const filterText = this.props.filterText.toLowerCase()
    let locations = this.props.locations

    if (filterText === 'wirschaffendas') {
      return filter(locations, (location) => !location.live)
    }

    locations = filter(locations, (location) => location.live)

    return filter(locations, (location) => location.name.toLowerCase().includes(filterText))
  }

  renderList (locations) {
    const groups = groupBy(locations, location => location.sortCategory)
    return transform(groups, (result, locations, key) => {
      result.push(<LocationParentEntry key={key} name={key}/>)
      result.push(locations.map((location, index) => <LocationEntry key={key + index}
                                                                    location={location}
                                                                    language={this.props.language}/>))
    }, [])
  }

  render () {
    return (
      <div>
        <div className={style.languageList}>
          {
            this.renderList(this.filter())
          }
        </div>
      </div>
    )
  }
}

export default LocationSelector
