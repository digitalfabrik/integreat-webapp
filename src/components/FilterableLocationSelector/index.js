import React from 'react'
import PropTypes from 'prop-types'

import Heading from './Heading'
import Search from 'components/Search/Search'
import LocationSelector from './LocationSelector'
import LocationModel from 'endpoints/models/LocationModel'

class FilterableLocationSelector extends React.Component {
  static propTypes = {
    locations: PropTypes.arrayOf(PropTypes.instanceOf(LocationModel)),
    language: PropTypes.string
  }

  constructor (props) {
    super(props)

    this.state = {
      filterText: ''
    }
  }

  render () {
    return (
      <div>
        <Heading/>
        <Search filterText={this.state.filterText}
                onFilterTextChange={(filterText) => this.setState({filterText: (filterText)})}/>
        <LocationSelector locations={this.props.locations} filterText={this.state.filterText} language={this.props.language}/>
      </div>
    )
  }
}

export default FilterableLocationSelector
