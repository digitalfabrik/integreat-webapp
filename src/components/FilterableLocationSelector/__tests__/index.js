import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import LocationModel from 'endpoints/models/LocationModel'
import FilterableLocationSelector from '../index'
import store from 'store'
import { mount } from 'enzyme'
import LocationSelector from 'components/FilterableLocationSelector/LocationSelector'
import Search from 'components/Search/Search'

jest.mock('react-i18next')

describe('FilterableLocationSelector', () => {
  const locations = [
    new LocationModel('City', 'city', true),
    new LocationModel('Other city', 'otherCity', true),
    new LocationModel('Not-live', 'nonlive', false),
    new LocationModel('Yet another city', 'yetanothercity', true)
  ]

  test('should render', () => {
    const component = renderer.create(
      <Provider store={store}>
        <FilterableLocationSelector
          language="de"
          locations={locations}/>
      </Provider>
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  test('should pass filterText to LocationSelector and filter', () => {
    const wrapper = mount(
      <Provider store={store}>
        <FilterableLocationSelector
          language="de"
          locations={locations}/>
      </Provider>
    )

    const search = wrapper.find(Search)
    search.prop('onFilterTextChange')('City')

    const selector = wrapper.find(LocationSelector)
    expect(selector.instance().filter()).toHaveLength(3)
  })
})
