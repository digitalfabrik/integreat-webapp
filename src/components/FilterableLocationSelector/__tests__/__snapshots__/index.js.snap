// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterableLocationSelector should render 1`] = `
<div>
  <div>
    <img
      className="logo"
      src="test-file-stub"
    />
    <h1
      className="caption"
    >
      Location
    </h1>
  </div>
  <div
    className={undefined}
  >
    <div
      className="search"
    >
      <span
        aria-hidden={true}
        className="fa fa-search searchIcon"
      />
      <input
        className="searchInput"
        defaultValue=""
        onChange={[Function]}
        placeholder="Search"
        type="text"
      />
    </div>
  </div>
  <div>
    <div
      className="languageList"
    >
      <div
        className="languageListParent"
      >
        C
      </div>
      <a
        className="languageListItem"
        href="/city/de"
        onClick={[Function]}
        target={undefined}
      >
        City
      </a>
      <div
        className="languageListParent"
      >
        O
      </div>
      <a
        className="languageListItem"
        href="/otherCity/de"
        onClick={[Function]}
        target={undefined}
      >
        Other city
      </a>
      <div
        className="languageListParent"
      >
        Y
      </div>
      <a
        className="languageListItem"
        href="/yetanothercity/de"
        onClick={[Function]}
        target={undefined}
      >
        Yet another city
      </a>
    </div>
  </div>
</div>
`;
